File: app/src/main/java/com/sdex/activityrunner/intent/converter/ExtrasSerializer.java
package com.sdex.activityrunner.intent.converter;

import android.text.TextUtils;
import com.sdex.activityrunner.intent.LaunchParamsExtra;
import java.util.ArrayList;
import java.util.List;

public final class ExtrasSerializer {

  private static final String DELIMITER_KEY_VALUE = "†";
  private static final String DELIMITER_EXTRA = "‡";

  public ExtrasSerializer() {
  }

  public String serialize(List<LaunchParamsExtra> list) {
    StringBuilder stringBuilder = new StringBuilder();
    for (int i = 0; i < list.size(); i++) {
      LaunchParamsExtra extra = list.get(i);
      stringBuilder.append(extra.getKey()).append(DELIMITER_KEY_VALUE)
        .append(extra.getValue()).append(DELIMITER_KEY_VALUE)
        .append(extra.getType()).append(DELIMITER_KEY_VALUE)
        .append(extra.isArray());
      if (i != list.size() - 1) {
        stringBuilder.append(DELIMITER_EXTRA);
      }
    }
    return stringBuilder.toString();
  }

  public ArrayList<LaunchParamsExtra> deserialize(String input) {
    if (TextUtils.isEmpty(input)) {
      return new ArrayList<>();
    }
    final String[] extras = input.split(DELIMITER_EXTRA);
    ArrayList<LaunchParamsExtra> output = new ArrayList<>(extras.length);
    for (String extra : extras) {
      final String[] values = extra.split(DELIMITER_KEY_VALUE);
      LaunchParamsExtra paramsExtra = new LaunchParamsExtra();
      paramsExtra.setKey(values[0]);
      paramsExtra.setValue(values[1]);
      paramsExtra.setType(Integer.parseInt(values[2]));
      paramsExtra.setArray(Boolean.parseBoolean(values[3]));
      output.add(paramsExtra);
    }
    return output;
  }
}

File: app/src/main/java/com/sdex/activityrunner/intent/converter/ExtrasSerializer.kt
package com.sdex.activityrunner.intent.converter

import com.sdex.activityrunner.intent.LaunchParamsExtra

class ExtrasSerializer {

    fun serialize(list: List<LaunchParamsExtra>): String {
        val stringBuilder = StringBuilder()
        for (i in list.indices) {
            val extra = list[i]
            stringBuilder.append(extra.key).append(DELIMITER_KEY_VALUE)
                .append(extra.value).append(DELIMITER_KEY_VALUE)
                .append(extra.type).append(DELIMITER_KEY_VALUE)
                .append(extra.isArray)
            if (i != list.size - 1) {
                stringBuilder.append(DELIMITER_EXTRA)
            }
        }
        return stringBuilder.toString()
    }

    fun deserialize(input: String?): ArrayList<LaunchParamsExtra> {
        if (input.isNullOrEmpty()) {
            return ArrayList()
        }
        val extras = input.split(DELIMITER_EXTRA.toRegex())
            .dropLastWhile { it.isEmpty() }
            .toTypedArray()
        val keyValueDelimiterRegex = DELIMITER_KEY_VALUE.toRegex()
        return ArrayList(
            extras.map { extra ->
                val values = extra.split(keyValueDelimiterRegex)
                    .dropLastWhile { it.isEmpty() }
                    .toTypedArray()
                LaunchParamsExtra(
                    key = values[0],
                    value = values[1],
                    type = Integer.parseInt(values[2]),
                    isArray = java.lang.Boolean.parseBoolean(values[3]),
                )
            },
        )
    }

    companion object {

        private const val DELIMITER_KEY_VALUE = "†"
        private const val DELIMITER_EXTRA = "‡"
    }
}
